# Exercícios de Regressão Linear Simples e Múltipla- Econometria 2020


# Capítulo 2 - Wooldridge 
### Exércicio Computacional 3

Use os dados do arquivo SLEEP75 de Biddle e Hamermesh (1990), para estudar se há uma compensação entre o tempo gasto dormindo por semana e o tempo gasto em um trabalho reumerado. Podemos usar qualquer variável como a variável dependente. Para materializar, estime o modelo: 

sleep= Bo + B1*totwrk + u 

em que sleep são os minutos dormidos à noite por semana e a totwrk é o total de minutos trabalhados durante a semana.

i) Registre seus resultados em uma equação junto com o número de observações e o R-quadrado. O que o intercepto desta equação significa? 

ii) Se totwrk aumentar 2 horas, quanto você estima que sleep cairá? Você acha que este é um efeito grande?



###### Imports...

In [1]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn import linear_model
import statsmodels.api as sm


###### Carregando os Dados...

In [2]:
dados=pd.read_stata('sleep75.dta.tga')
dados.columns


Index(['age', 'black', 'case', 'clerical', 'construc', 'educ', 'earns74',
       'gdhlth', 'inlf', 'leis1', 'leis2', 'leis3', 'smsa', 'lhrwage',
       'lothinc', 'male', 'marr', 'prot', 'rlxall', 'selfe', 'sleep',
       'slpnaps', 'south', 'spsepay', 'spwrk75', 'totwrk', 'union', 'worknrm',
       'workscnd', 'exper', 'yngkid', 'yrsmarr', 'hrwage', 'agesq'],
      dtype='object')

In [3]:
dados

age  black   case  clerical  construc  educ  earns74  gdhlth  inlf  \
0    32.0    0.0    1.0  0.000000  0.000000  12.0      0.0     0.0   1.0   
1    31.0    0.0    2.0  0.000000  0.000000  14.0   9500.0     1.0   1.0   
2    44.0    0.0    3.0  0.000000  0.000000  17.0  42500.0     1.0   1.0   
3    30.0    0.0    4.0  0.000000  0.000000  12.0  42500.0     1.0   1.0   
4    64.0    0.0    5.0  0.000000  0.000000  14.0   2500.0     1.0   1.0   
..    ...    ...    ...       ...       ...   ...      ...     ...   ...   
701  45.0    0.0  702.0  0.182331  0.030075  12.0   5500.0     1.0   0.0   
702  34.0    0.0  703.0  0.182331  0.030075  10.0   2500.0     0.0   0.0   
703  37.0    0.0  704.0  0.182331  0.030075  12.0   3500.0     1.0   0.0   
704  54.0    0.0  705.0  0.182331  0.030075  17.0  32500.0     1.0   0.0   
705  30.0    0.0  706.0  0.182331  0.030075  16.0   6750.0     1.0   0.0   

      leis1  ...  spwrk75  totwrk  union  worknrm  workscnd  exper  yngkid  \
0    3529.0  ...      0.0  3438.0    0.0   3438.0       0.0   14.0     0.0   
1    2140.0  ...      0.0  5020.0    0.0   5020.0       0.0   11.0     0.0   
2    4595.0  ...      1.0  2815.0    0.0   2815.0       0.0   21.0     0.0   
3    3211.0  ...      1.0  3786.0    0.0   3786.0       0.0   12.0     0.0   
4    4052.0  ...      1.0  2580.0    0.0   2580.0       0.0   44.0     0.0   
..      ...  ...      ...     ...    ...      ...       ...    ...     ...   
701  5069.0  ...      1.0  2026.0    0.0   2026.0       0.0   27.0     0.0   
702  5885.0  ...      0.0   675.0    1.0    465.0     210.0   18.0     0.0   
703  4719.0  ...      1.0  1851.0    0.0   1851.0       0.0   19.0     0.0   
704  5149.0  ...      1.0  1961.0    1.0   1481.0     480.0   31.0     0.0   
705  4747.0  ...      0.0  2363.0    0.0   2363.0       0.0    8.0     1.0   

     yrsmarr     hrwage   agesq  
0       13.0   7.070004  1024.0  
1        0.0   1.429999   961.0  
2        0.0  20.530001  1936.0  
3       12.0   9.619998   900.0  
4       33.0   2.750000  4096.0  
..       ...        ...     ...  
701     18.0        NaN  2025.0  
702      4.0        NaN  1156.0  
703     17.0        NaN  1369.0  
704     22.0        NaN  2916.0  
705      9.0        NaN   900.0  

[706 rows x 34 columns]

###### Carregando as variáveis para a regressão...

In [4]:
y=dados['sleep']
X=dados['totwrk']

###### Estimando a regressão... 

In [5]:
# é necessário adicionar uma constante a matriz X
X_sm = sm.add_constant(X)
# OLS vem de Ordinary Least Squares e o método fit irá treinar o modelo
results = sm.OLS(y, X_sm).fit()
# mostrando as estatísticas do modelo
results.summary()
# mostrando as previsões para o mesmo conjunto passado
results.predict(X_sm)
results.summary()

/opt/anaconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sleep   R-squared:                       0.103
Model:                            OLS   Adj. R-squared:                  0.102
Method:                 Least Squares   F-statistic:                     81.09
Date:                Wed, 18 Mar 2020   Prob (F-statistic):           1.99e-18
Time:                        11:42:34   Log-Likelihood:                -5267.1
No. Observations:                 706   AIC:                         1.054e+04
Df Residuals:                     704   BIC:                         1.055e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3586.3770     38.912     92.165      0.000    3509.979    3662.775
totwrk        -0.1507      0.017     -9.005      0.000      -0.184      -0.118
==============================================================================
Omnibus:                       68.651   Durbin-Watson:                   1.955
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              192.044
Skew:                          -0.483   Prob(JB):                     1.99e-42
Kurtosis:                       5.365   Cond. No.                     5.71e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.71e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### item i)

In [6]:
3586/420

8.538095238095238

In [7]:
print('A equação é estimada é dada por: sleep = 3586.3370 -0.1507*totwrk')
print('R-quadrado:0.103')
print('Amostra: 706 observações')

print('Podemos interpretar o intercepto como: um indivíduo que não trabalha dorme 3586 minutos por semana, isto é aproximadamente 8,5 horas por noite')

A equação é estimada é dada por: sleep = 3586.3370 -0.1507*totwrk
R-quadrado:0.103
Amostra: 706 observações
Podemos interpretar o intercepto como: um indivíduo que não trabalha dorme 3586 minutos por semana, isto é aproximadamente 8,5 horas por noite


### item ii)

In [8]:
minutos_totais=60*2
variacao_sleep= 0.1507*minutos_totais
print('Estimo que o indivíduo dormirá '+ str(variacao_sleep) + ' minutos a menos')

Estimo que o indivíduo dormirá 18.084 minutos a menos


# Capítulo 2 - Wooldridge 
### Exércicio Computacional 6

Usamos os dados do arquivo  MEAP93 no exemplo 2.12. Agora, queremos explorar a relação entre a taxa de aprovação em matemática (math10) e os gastos por estudante (expend)

i) Você acha que cada dólar adicional gasto tem o mesmo efeito sobre a taxa de aprovação ou um efeito decrescente seria mais razoável? Explique.

ii) No modelo populacional 

math10=Bo + B1*log(expend) + u argumente que B1/10 é a porcentagem de alteração em math10 dado um aumento de 10% em gasto. 

iii) Use os dados do arquivo MEAP93 para estimar o modelo ii. Descreva a equação estimada da forma usual, incluindo o tamanho da amostra e o R-quadrado.

iv) Quão grande é o efeito de gastos estimado? Em outras palavras, se os gastos aumentarem 10%, qual será o aumento percentual estimado em math10? 

v) Alguns podem se preocupar com o fato de que a análise de regressão pode produzir valores ajustados para math10 maiores do que 100. Por que isso não é tão preocupante neste conjunto de dados?




#### Carregando os dados...

In [9]:
dados=pd.read_stata('meap93.dta.tga')
dados

lnchprg   enroll       staff  expend   salary  benefits   droprate  \
0     1.400000   1862.0  112.599998  5765.0  37498.0    7420.0   2.900000   
1     2.300000  11355.0  101.199997  6601.0  48722.0   10370.0   1.300000   
2     2.700000   7685.0  114.000000  6834.0  44541.0    7313.0   3.500000   
3     3.400000   1148.0   85.400002  3586.0  31566.0    5989.0   3.600000   
4     3.400000   1572.0   96.099998  3847.0  29781.0    5545.0   0.000000   
..         ...      ...         ...     ...      ...       ...        ...   
403  60.900002   1689.0   78.400002  5380.0  41856.0    9468.0  35.799999   
404  61.099998   2631.0  100.699997  5088.0  35020.0    7655.0  17.500000   
405  73.800003   4024.0   85.199997  3942.0  32678.0    6220.0   8.700000   
406  74.400002    850.0  122.400002  5029.0  29210.0    6513.0   1.600000   
407  79.500000   7585.0   93.300003  4529.0  33645.0    7682.0  22.000000   

       gradrate     math10      sci11  totcomp  ltotcomp   lexpend   lenroll  \
0     89.199997  56.400002  67.900002  44918.0  10.71259  8.659560  7.529407   
1     91.400002  42.700001  65.300003  59092.0  10.98685  8.794976  9.337414   
2     91.400002  43.799999  54.299999  51854.0  10.85619  8.829665  8.947025   
3     86.599998  25.299999  60.000000  37555.0  10.53356  8.184793  7.045776   
4    100.000000  15.300000  65.800003  35326.0  10.47237  8.255049  7.360104   
..          ...        ...        ...      ...       ...       ...       ...   
403   23.500000   7.000000  11.200000  51324.0  10.84591  8.590444  7.431892   
404   61.099998   8.100000  22.700001  42675.0  10.66137  8.534640  7.875119   
405   83.300003  11.500000  26.799999  38898.0  10.56870  8.279444  8.300032   
406   91.900002   9.500000   9.700000  35723.0  10.48355  8.522977  6.745236   
407   40.599998   2.700000   7.200000  41327.0  10.62927  8.418257  8.933928   

       lstaff    bensal   lsalary  
0    4.723842  0.197877  10.53204  
1    4.617099  0.212840  10.79389  
2    4.736198  0.164186  10.70417  
3    4.447346  0.189729  10.35984  
4    4.565389  0.186192  10.30163  
..        ...       ...       ...  
403  4.361824  0.226204  10.64199  
404  4.612146  0.218589  10.46367  
405  4.445002  0.190342  10.39446  
406  4.807294  0.222972  10.28227  
407  4.535820  0.228325  10.42362  

[408 rows x 17 columns]

### i)

O efeito de cada dolar adicional gasto é decrescente, uma vez que em escolas com maiores gastos já se espera professores de alta qualidade, bons materiais etc 

### ii)

Δmath10=β1Δlog(expend)≈(β1 /100)(%Δexpend), o que evidencia o pedido. 

### iii)

#### Carregando as variáveis...

In [10]:
y=dados['math10']
X=dados['lexpend']

#### Estimando a regressão...

In [11]:
# é necessário adicionar uma constante a matriz X
X_sm = sm.add_constant(X)
# OLS vem de Ordinary Least Squares e o método fit irá treinar o modelo
results = sm.OLS(y, X_sm).fit()
# mostrando as estatísticas do modelo
results.summary()
# mostrando as previsões para o mesmo conjunto passado
results.predict(X_sm)
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 math10   R-squared:                       0.030
Model:                            OLS   Adj. R-squared:                  0.027
Method:                 Least Squares   F-statistic:                     12.41
Date:                Wed, 18 Mar 2020   Prob (F-statistic):           0.000475
Time:                        18:25:03   Log-Likelihood:                -1531.4
No. Observations:                 408   AIC:                             3067.
Df Residuals:                     406   BIC:                             3075.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -69.3410     26.530     -2.614      0.009    -121.495     -17.187
lexpend       11.1644      3.169      3.523      0.000       4.935      17.394
==============================================================================
Omnibus:                       28.397   Durbin-Watson:                   1.615
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               34.609
Skew:                           0.591   Prob(JB):                     3.05e-08
Kurtosis:                       3.800   Cond. No.                         440.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [14]:
print('-69.3410 + 11.1644(lexpend)')
print('N=408; R-quadrado=0.030')

-69.3410 + 11.1644(lexpend)
N=408; R-quadrado=0.030


### iv)

Se os gastos aumentarem 10%, math10 aumenta 1.1% em sua pontuação

### v)

#### Ordenando as variáveis do DataFrame a partir da coluna 'math10' com o uso da função sort_values 

In [16]:
ordenacao=dados.sort_values(by='math10',ascending=False).reset_index()
ordenacao[0:6]

index    lnchprg  enroll       staff  expend   salary  benefits  droprate  \
0    320  35.400002   226.0  162.800003  6444.0  26653.0    8191.0       0.0   
1     24   6.900000  2119.0   89.199997  4018.0  33345.0    6373.0       3.1   
2    235  27.200001  1234.0   90.300003  3808.0  30639.0    6148.0       0.5   
3      0   1.400000  1862.0  112.599998  5765.0  37498.0    7420.0       2.9   
4     23   6.800000   650.0   97.599998  4403.0  34194.0    6833.0       0.0   
5    142  18.400000   222.0  122.500000  5365.0  32249.0    6275.0      11.3   

     gradrate     math10      sci11  totcomp  ltotcomp   lexpend   lenroll  \
0  100.000000  66.699997  64.900002  34844.0  10.45864  8.770905  5.420535   
1   87.199997  62.400002  66.699997  39718.0  10.58956  8.298539  7.658700   
2   97.699997  60.599998  41.200001  36787.0  10.51290  8.244860  7.118016   
3   89.199997  56.400002  67.900002  44918.0  10.71259  8.659560  7.529407   
4  100.000000  54.799999  72.800003  41027.0  10.62199  8.390041  6.476973   
5   61.000000  53.799999  39.599998  38524.0  10.55904  8.587651  5.402678   

     lstaff    bensal   lsalary  
0  5.092523  0.307320  10.19066  
1  4.490881  0.191123  10.41466  
2  4.503138  0.200659  10.33003  
3  4.723842  0.197877  10.53204  
4  4.580877  0.199830  10.43981  
5  4.808111  0.194580  10.38124

Note que o maior valor de math10 não atinge 67 o que é distante de 100.

# Capítulo 3 - Wooldridge 
### Exércicio Computacional 7

Use os dados do arquivo MEAP93 para responder a essa questão.

i) Estime o modelo 

math10=Bo + B1*log(expend) + B2*lnchprg + u e registre os resultados na forma usual, incluindo o tamanho da amostra e o R-quadrado. Os sinais do coeficiente de inclinação são os que você esperava? Explique.

ii) Como você interpreta o intercepto que estimou no item (i)? Em especial, faz sentido definir duas variáveis explicativas como zero? [Lembre-se que log(1)=0]. 

iii) Agora faça a regressão simples de math sobre log(expend) e compare o coeficiente de inclinação com aquele obtido no item (i). O efeito de gastos estimado agora é maior ou menor do que no item i)?

iv) Encontre a correlação entre lexpend=log(expend) e lnchprg. O sinal da relação faz sentido para você? 

v) Use o item iv) para explicar suas descobertas no item (iii). 

### item i)

Note que este exercício usa a mesma base de dados do exercício anterior, portanto não atualizaremos a base de dados. 

#### Carregando as variáveis e estimando a regressão...

In [17]:
from pandas import DataFrame
from sklearn import linear_model
import statsmodels.api as sm



X = dados[['lexpend','lnchprg']] # here we have 2 variables for multiple regression. If you just want to use one variable for simple linear regression, then use X = df['Interest_Rate'] for example.Alternatively, you may add additional variables within the brackets
Y = dados['math10']
 
# with sklearn
regr = linear_model.LinearRegression()
regr.fit(X, Y)

print('Intercept: \n', regr.intercept_)
print('Coefficients: \n', regr.coef_)

# prediction with sklearn
#New_Interest_Rate = 2.75
#New_Unemployment_Rate = 5.3
#print ('Predicted Stock Index Price: \n', regr.predict([[New_Interest_Rate ,New_Unemployment_Rate]]))

# with statsmodels
X = sm.add_constant(X) # adding a constant
 
model = sm.OLS(Y, X).fit()
predictions = model.predict(X) 
 
print_model = model.summary()
print(print_model)

Intercept: 
 -20.360691
Coefficients: 
 [ 6.2296824  -0.30458528]
                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.180
Model:                            OLS   Adj. R-squared:                  0.176
Method:                 Least Squares   F-statistic:                     44.43
Date:                Wed, 18 Mar 2020   Prob (F-statistic):           3.59e-18
Time:                        18:38:36   Log-Likelihood:                -1497.1
No. Observations:                 408   AIC:                             3000.
Df Residuals:                     405   BIC:                             3012.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------

/opt/anaconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [19]:
print('-20.360691 + 6.2296824(lexpend) -0.30458528(lnchprg)')
print('N:408  R-quadrado: 0.180')
print('Os sinais dos coeficientes de inclinação implicam que mais gastos aumenta a taxa de aprovação dado lnchprg fixo e maior taxa de pobreza (lnchprg) diminui a taxa de aprovação dado o gasto fixo. Era o esperado')

-20.360691 + 6.2296824(lexpend) -0.30458528(lnchprg)
N:408  R-quadrado: 0.180
Os sinais dos coeficientes de inclinação implicam que mais gastos aumenta a taxa de aprovação dado lnchprg fixo e maior taxa de pobreza (lnchprg) diminui a taxa de aprovação dado o gasto fixo. Era o esperado


### ii)

O intercepto estimado prediz o valor da variável dependende quando todos os regressores são zero. Fazendo lnchprg=0 faz sentido uma vez que existem escolas com baixa taxa de pobreza. Fazendo lexpend=0 não faz sentido, uma vez que é o mesmo que fazer expend=1, mas essa variáve; é medida em dólares por estudante. Deste modo, -20 é uma previsão sem sentido. 

### iii)

In [21]:
print('-69.3410 + 11.1644(lexpend)')
print('N=408; R-quadrado=0.030')

-69.3410 + 11.1644(lexpend)
N=408; R-quadrado=0.030


Este item é equivalente ao item iii) do exercício anterior. Nota-se portanto que o efeito do gasto é bem maior em comparação a regressão múltiplica (quase o dobro) 

### iv)

#### Matriz de Correlação

In [27]:
correlacao=pd.DataFrame()
correlacao['lexpend']=dados['lexpend']
correlacao['lnchprg']=dados['lnchprg']
rets3 = correlacao.pct_change()  
corr = rets3.corr()
corr


lexpend   lnchprg
lexpend  1.000000 -0.107143
lnchprg -0.107143  1.000000

Deste modo temos que a correlação é de -0.107143, evidenciando que escolas com estudantes mais pobres gastam menos por estudante

### v)

Como a correlacão é negativa, observamos que ao não se considerar o efeito da taxa de pobreza, os efeitos dos gastos adicionais são superestimados.

# Capítulo 3 - Wooldridge 
### Exércicio Computacional 8

Use os dados do arquivo DISCRIM para responder a essa questão. São dados de preços de vários itens em restaurantes de fast food e características da população dividida por CEP em Nova Jersey e Pensilvânia. A ideia é ver se os restaurantes cobram preços mais altos em áreas com maior concentração de negros. 

i) Encontre os valores médios da proporção de negros (prpblck) e renda (income) na amostra, além de seus desvios padrão. Quais são as unidades de medida prpblck e income? 

ii) Considere um modelo para explicar o preço do refrigerante, psoda, em termos de proporção da população que é negra e com renda mediana: 

psoda= Bo + B1*prpblck + B2*income + u 

Estime o modelo por MQO e registre os resultados em forma de equação, incluindo o tamanho da amostra e o R-quadrado. (Não use notação científica ao relatar as estimativas. Interprete o coeficiente sobre prpblck. Você acha que ele é economicamente grande? 

iii) Compare a estimativa do item (ii) com a estimativa de regressão simples de psoda sobre prpblck. O efeito de discriminação é maior ou menor quando se controla a renda? 

iv) Um modelo de elasticidade-preço constante em relação à renda pode ser mais adequado. Registre as estimativas do modelo 

log(psoda)= Bo + B1*prpblck + B2*log(income) + u 

Se prpblck aumenta 0,2 (20 pontos percentuais), qual é a alteração percentual estimada em psoda? (Dica: A resposta é 2.xx, você encontrar o 'xx'.) 

v) Agora adicione a variável prppov à regressão do item (iv). O que acontece com B(prpblck)? 

vi) Encontre a correlação entre log(income) e prppov. É aproximadamente o que você espera? 

vii) Avalie a seguinte afirmação: "Como log(income) e prppob são altamente correlacionadas, elas não têm por que estar na mesma regressão". 

#### Carregando os dados...

In [136]:
dados=pd.read_stata('discrim.dta')
dados

psoda  pfries  pentree  wagest  nmgrs  nregs  hrsopen   emp  psoda2  \
0     1.12    1.06     1.02    4.25    3.0    5.0     16.0  27.5    1.11   
1     1.06    0.91     0.95    4.75    3.0    3.0     16.5  21.5    1.05   
2     1.06    0.91     0.98    4.25    3.0    5.0     18.0  30.0    1.05   
3     1.12    1.02     1.06    5.00    4.0    5.0     16.0  27.5    1.15   
4     1.12     NaN     0.49    5.00    3.0    3.0     16.0   5.0    1.04   
..     ...     ...      ...     ...    ...    ...      ...   ...     ...   
405   1.11    1.05     0.94    4.50    3.0    2.0     12.5  10.0    1.11   
406   0.95    0.74     2.33    4.75    3.0    4.0     10.0   8.0    0.95   
407   0.97    0.84     0.91    4.25    4.0    3.0     18.0  35.0    0.97   
408   0.97    0.86     0.89    4.75    4.0    3.0     17.0  18.5    0.97   
409   1.02    0.89     0.90    4.50    3.0    3.0     15.0  29.0    1.02   

     pfries2  ...  county    lpsoda   lpfries    lhseval    lincome  ldensity  \
0       1.11  ...      18  0.113329  0.058269  11.906993  10.704008  8.301521   
1       0.89  ...      18  0.058269 -0.094311  11.906993  10.704008  8.301521   
2       0.94  ...      12  0.058269 -0.094311  12.038836  10.625319  9.341369   
3       1.05  ...      10  0.113329  0.019803  12.052921  10.827071  9.029418   
4       1.01  ...      10  0.113329       NaN  12.425610  11.188399  6.579251   
..       ...  ...     ...       ...       ...        ...        ...       ...   
405     1.11  ...      20  0.104360  0.048790  11.946006  10.908375  7.253400   
406     0.84  ...      20 -0.051293 -0.301105  11.494761  10.611154  8.368465   
407     0.86  ...      20 -0.030459 -0.174353  11.494761  10.611154  8.368465   
408     0.84  ...      20 -0.030459 -0.150823  11.494761  10.611154  8.368465   
409     0.94  ...       3  0.019803 -0.116534  11.684355  10.531483  6.114567   

     NJ  BK  KFC  RR  
0     1   0    0   1  
1     1   1    0   0  
2     1   1    0   0  
3     1   0    0   1  
4     1   1    0   0  
..   ..  ..  ...  ..  
405   0   0    0   1  
406   0   0    1   0  
407   0   1    0   0  
408   0   1    0   0  
409   0   1    0   0  

[410 rows x 37 columns]

### i)

In [46]:
dados['prpblck'].mean()

0.11348654329776764

In [47]:
dados['income'].mean()

47053.7890625

In [48]:
dados['prpblck'].std()

0.18241645395755768

In [49]:
dados['income'].std()

13179.2880859375

### ii)

#### Nesta base de dados precisamos de alguns ajustes uma vez que algumas linhas estavam com valores considerados vazios (null). Deste modo limpamos a base de dados com a funcão drop.na() e depois certificamos que o número de observacões das variáveis explicada e explicativas eram os mesmos (função len()) 

In [123]:
X = dados[['prpblck','income']] 
Y = dados['psoda']

In [124]:
X=X.dropna()
Y=Y.dropna()


In [125]:
X=X[:len(Y)]

### Estimando a regressão...

In [110]:

# with sklearn
regr = linear_model.LinearRegression()
regr.fit(X, Y)

print('Intercept: \n', regr.intercept_)
print('Coefficients: \n', regr.coef_)

X = sm.add_constant(X) # adding a constant
Y=list(Y)
model = sm.OLS(Y, X).fit()
predictions = model.predict(X) 

print_model = model.summary()
print(print_model)


Intercept: 
 0.9734553803894059
Coefficients: 
 [0.00000000e+00 1.12653064e-01 1.24169787e-06]
                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.051
Model:                            OLS   Adj. R-squared:                  0.046
Method:                 Least Squares   F-statistic:                     10.75
Date:                Thu, 19 Mar 2020   Prob (F-statistic):           2.84e-05
Time:                        08:39:54   Log-Likelihood:                 414.53
No. Observations:                 402   AIC:                            -823.1
Df Residuals:                     399   BIC:                            -811.1
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------

/opt/anaconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [114]:
print('psoda = 0.97345 + 0.1127(prpblck) + 0.1127(income)')
print('n=402, R-quadrado=.051')

psoda = 0.97345 + 0.1127(prpblck) + 0.1127(income)
n=402, R-quadrado=.051


Deste modo, se prpblck aumentar 10%, o preço da soda é estimado para aumentar .0112 dolares, ou 1.12 centavos.

### iii)

#### Carregando as variáveis...

In [126]:
Y=Y
X=X['prpblck']

In [133]:
X=list(X)
y=Y

#### Estimando a regressão...

In [134]:
# é necessário adicionar uma constante a matriz X
X_sm = sm.add_constant(X)
# OLS vem de Ordinary Least Squares e o método fit irá treinar o modelo
results = sm.OLS(y, X_sm).fit()
# mostrando as estatísticas do modelo
results.summary()
# mostrando as previsões para o mesmo conjunto passado
results.predict(X_sm)
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.023
Model:                            OLS   Adj. R-squared:                  0.021
Method:                 Least Squares   F-statistic:                     9.495
Date:                Thu, 19 Mar 2020   Prob (F-statistic):            0.00220
Time:                        08:51:29   Log-Likelihood:                 408.70
No. Observations:                 402   AIC:                            -813.4
Df Residuals:                     400   BIC:                            -805.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.0364      0.005    200.759      0.000       1.026       1.047
x1             0.0735      0.024      3.081      0.002       0.027       0.120
==============================================================================
Omnibus:                       30.430   Durbin-Watson:                   1.651
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               65.920
Skew:                           0.407   Prob(JB):                     4.85e-15
Kurtosis:                       4.809   Cond. No.                         5.52
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Resposta Final:
Nota-se que a regressão simples tem o coeficiente de prpblck estimado em .0735, isto é, menor em comparação à regressão mútipla (.112). Isto ocorre pois prpblck e renda são negativamente correlacionados e renda tem um coeficiente positivo na regressão múltipla.

### iv)

#### Carregando os dados...

In [159]:
Y=dados['lpsoda']
X=dados[['prpblck','lincome']]
Y=Y.dropna()
X=X.dropna()
X=X[:len(Y)]
Y=list(Y)
X

prpblck    lincome
0    0.171154  10.704008
1    0.171154  10.704008
2    0.047360  10.625319
3    0.052839  10.827071
4    0.034480  11.188399
..        ...        ...
398  0.033835  11.331152
399  0.022458  10.840071
400  0.022458  10.840071
401  0.022458  10.840071
402  0.034672  10.652495

[402 rows x 2 columns]

#### Estimando a regressão

In [160]:
# with sklearn
regr = linear_model.LinearRegression()
regr.fit(X, Y)

print('Intercept: \n', regr.intercept_)
print('Coefficients: \n', regr.coef_)

X = sm.add_constant(X) # adding a constant
Y=list(Y)
model = sm.OLS(Y, X).fit()
predictions = model.predict(X) 

print_model = model.summary()
print(print_model)



Intercept: 
 -0.56089884
Coefficients: 
 [0.11441682 0.05485626]
                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.050
Model:                            OLS   Adj. R-squared:                  0.045
Method:                 Least Squares   F-statistic:                     10.51
Date:                Thu, 19 Mar 2020   Prob (F-statistic):           3.58e-05
Time:                        09:02:50   Log-Likelihood:                 432.66
No. Observations:                 402   AIC:                            -859.3
Df Residuals:                     399   BIC:                            -847.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------

/opt/anaconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [162]:
print('lpsoda = -0.569 + 0.1144(prpblck) + 0.0549(lincome)')
print('n=402, R-quadrado=.050')

lpsoda = -0.569 + 0.1144(prpblck) + 0.0549(lincome)


In [166]:
print('Deste modo se prpblck aumenta .20, log(psoda) é estimado para aumentar ' + str(0.20*0.1144) + ' ou ' + str(100*0.20*0.1144) + '%')

Deste modo se prpblck aumenta .20, log(psoda) é estimado para aumentar 0.02288 ou 2.2880000000000003%


### v)

#### Carregando os dados...

In [167]:
Y=dados['lpsoda']
X=dados[['prpblck','lincome','prppov']]
Y=Y.dropna()
X=X.dropna()
X=X[:len(Y)]
Y=list(Y)
X

prpblck    lincome    prppov
0    0.171154  10.704008  0.036579
1    0.171154  10.704008  0.036579
2    0.047360  10.625319  0.087907
3    0.052839  10.827071  0.059123
4    0.034480  11.188399  0.025415
..        ...        ...       ...
398  0.033835  11.331152  0.036649
399  0.022458  10.840071  0.027570
400  0.022458  10.840071  0.027570
401  0.022458  10.840071  0.027570
402  0.034672  10.652495  0.007743

[402 rows x 3 columns]

#### Estimando a regressão...

In [168]:
# with sklearn
regr = linear_model.LinearRegression()
regr.fit(X, Y)

print('Intercept: \n', regr.intercept_)
print('Coefficients: \n', regr.coef_)

X = sm.add_constant(X) # adding a constant
Y=list(Y)
model = sm.OLS(Y, X).fit()
predictions = model.predict(X) 

print_model = model.summary()
print(print_model)

Intercept: 
 -1.3686599
Coefficients: 
 [0.0558176  0.12775162 0.45953408]
                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.077
Model:                            OLS   Adj. R-squared:                  0.070
Method:                 Least Squares   F-statistic:                     11.13
Date:                Thu, 19 Mar 2020   Prob (F-statistic):           4.99e-07
Time:                        09:10:38   Log-Likelihood:                 438.53
No. Observations:                 402   AIC:                            -869.1
Df Residuals:                     398   BIC:                            -853.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------

/opt/anaconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [171]:
print('lsoda = -1.368 + 0.00558(prpblck) + 0.1278(lincome) + 0.4595(prppov)')
print('n=402, R-quadrado=0.077')
print('Note que prpblck sai de 0.11 para 0.0055')

lsoda = -1.368 + 0.00558(prpblck) + 0.1278(lincome) + 0.4595(prppov)
n=402, R-quadrado=0.077
Note que prpblck sai de 0.11 para 0.0055


### vi)

#### Construindo a matriz de correlação entre as duas variáveis..

In [172]:
correlacao=pd.DataFrame()
correlacao['prpblck']=dados['prpblck']
correlacao['lincome']=dados['lincome']
rets3 = correlacao.pct_change()  
corr = rets3.corr()
corr



prpblck   lincome
prpblck  1.000000 -0.233504
lincome -0.233504  1.000000

Note que a correlação é -0.233, o que faz sentido, uma vez que a taxa de pobreza é determinada pela renda, mas não diretamente em termos da renda média

### vii)

Não há argumentos para justificar alta correlação, mas estamos tentando determinar se há discriminação de preços para negros. Para isolarmos o efeito de discriminação totalmente precisariamos de mais medidas de renda do que as expostas. A inclusão das duas variáveis faz sentido.